Feature/graphql server settings

schroda opened the pull request:
is rebased onto #617

schroda committed:
Add query for server settings

schroda committed:
Add mutation for server settings

schroda committed:
Rename "WebUI" enum to "WebUIFlavor"

schroda committed:
Add "uiName" to WebUI enum

schroda committed:
Add "Custom" WebUI to enum

schroda committed:
Add "WebUIInterface" enum

schroda committed:
Add mutation to reset the server settings

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/graphql/types/SettingsType.kt:
```diff
@@ -0,0 +1,208 @@
+    val localSourcePath: String?
+}
+
+data class PartialSettingsType(
```
I see that these are nullable, but they don't have a default value, does this allow partial editing or does the user have to set the values as null manually?

schroda replied:
it does, I guess because it's a property of a required object parameter?

Syer10 replied:
Then its good 👍

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/graphql/mutations/SettingsMutation.kt:
```diff
@@ -0,0 +1,89 @@
+        val (clientMutationId) = input
+
+        GlobalConfigManager.resetUserConfig()
+        val defaultServerConfig = ServerConfig({ GlobalConfigManager.config.getConfig(SERVER_CONFIG_MODULE_NAME) })
```
This resets the running server? All the MutableStateFlows update properly?

schroda replied:
yes, this is what happens:

 - `GlobalConfigManager.resetUserConfig()` resets the config file
 - `val defaultServerConfig = ...` creates a temp ServerConfig instance to access the variables without having to manually retrieve them from the config
 - `updateSettings(settings)` updates the flows

and I have to add a guard to prevent updating the config file in case the value that should be set is the same as the current value in the config
since currently the file gets reset and then every flow value change updates the file again...

- [x] add guard to config value updates

schroda replied:
added guard

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/graphql/mutations/SettingsMutation.kt:
```diff
@@ -0,0 +1,89 @@
+        return SetSettingsPayload(clientMutationId, SettingsType())
+    }
+
+    data class ResetSettingsInput(val clientMutationId: String?)
```
Add default null

schroda replied:
done

schroda committed:
Only update the config in case the value changed

In case the value of the config is already the same as the new value of the state flow, it is not necessary to update the config file
